herculens/LensImage/lens_image_multiplane.py

```diff
@@ -0,0 +1,314 @@
+                    **self.MPLightModel.light_models[i].pixel_grid_settings
+                )
+                self.MPLightModel.light_models[i].set_pixel_grid(pixel_grid, self.Grid.pixel_area)
+        if source_arc_masks is None:
+            self.source_arc_masks = np.stack(
+                [np.ones(self.Grid.num_pixel_axes)] * self.MPLightModel.number_light_planes
+            )
+        else:
+            self.source_arc_masks = np.stack([
+                np.ones(self.Grid.num_pixel_axes) if m is None else m for m in source_arc_masks
+            ])
+
+        self._src_adaptive_grid = self.MPLightModel.pixel_is_adaptive
+
+        if kwargs_numerics is None:
+            kwargs_numerics = {}
+        self.ImageNumerics = Numerics(pixel_grid=self.Grid, psf=self.PSF, **kwargs_numerics)
+
+        ssf = self.ImageNumerics.grid_supersampling_factor
+
+        # get masks in super sampled space
+        s_ones = np.ones([ssf, ssf])
+        self.source_arc_masks_ss = np.stack([
+            np.kron(m, s_ones) for m in self.source_arc_masks
+        ])
+        # flatten the super sampled masks
+        self._source_arc_masks_flat = self.source_arc_masks_ss.reshape(
+            self.MPLightModel.number_light_planes,
+            -1
+        )
+        # get the (flattened) outline of the super sampled masks
+        # these boundaries are used to define the extent of pixelated grids
+        self._source_arc_masks_flat_bool = np.stack([
+            (m - scipy.ndimage.binary_erosion(m))
+            for m in self.source_arc_masks_ss
+        ]).reshape(
+            self.MPLightModel.number_light_planes,
+            -1
+        ).astype(bool)
+
+    def k_extend(self, k, amount):
+        if k is None:
+            return jnp.arange(amount)
+        else:
+            return jnp.array(k)
+
+    @partial(jax.jit, static_argnums=(0, 4, 5, 6, 7, 8))
+    def model(
+        self,
+        eta_flat=None,
+        kwargs_mass=None,
+        kwargs_light=None,
+        supersampled=False,
+        k_mass=None,
+        k_light=None,
+        k_planes=None,
+        return_pixel_scale=False
+    ):
+        ra_grid_img, dec_grid_img = self.ImageNumerics.coordinates_evaluate
+
+        # pixel grid positions on each mass plane (including the lens plane)
+        ra_grid_planes, dec_grid_planes = self.MPMassModel.ray_shooting(
+            ra_grid_img,
+            dec_grid_img,
+            eta_flat,
+            kwargs_mass,
+            k=k_mass
+        )
+        # (masked) light contribution from each plane
+        pixels_x_coord, pixels_y_coord, _ = self.adapt_source_coordinates(
+            ra_grid_planes,
+            dec_grid_planes
+        )
+        light_planes = self.MPLightModel.surface_brightness(
+            ra_grid_planes,
+            dec_grid_planes,
+            kwargs_light,
+            pixels_x_coord,
+            pixels_y_coord,
+            k=k_light,
+        ) * self._source_arc_masks_flat
```

Comment:
@CKrawczyk I think the multiplication by the arc masks should be performed after convolution, to avoid introducing artifacts in the model at the edge of the masks.

Reply:
We did it this way because the convolution happens once on the final image rather than each slice independently.  To get around the artifact a "soft mask" can be used (gaussian blur the mask so it is a smooth gradient down to zero at the edge).